ParU seems to require "ENABLE_CUDA" to match CHOLMOD, KLU, etc.

Hi @mmuetzel :  I tried the recent build system and hit a small snag in ParU (and fixed it, I think).  

CHOLMOD was compiled with the GPU but ParU doesn't have an "ENABLE_CUDA" option, so it failed because CUDA::cublas wasn't found.  This is likely because ParU was linking against just -lcholmod, not -lcholmod_cuda.  Now that the cholmod and cholmod_cuda libraries are merged, ParU complains about CUDA::cublas not being found.  I fixed it by adding "ENABLE_CUDA" to the ParU CMakeLists.txt file (see https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/eb2809103b76ef907807a7c63c57038d26e5fb3a ) but that might not be the best fix.

Here's how it failed.  I'm using the top-level 'make' and thus it become "cd ParU/build ; cmake .. ; etc", not the root CMakeLists.txt file (that might work because ENABLE_CUDA would likely be enabled for all of SuiteSparse in that case).

This might not be an issue, since I've now fixed it, but I might not have fixed it in the most elegant way.  I see that KLU has an "ENABLE_CUDA" line in it as well, so perhaps this is the right fix since KLU relies on CHOLMOD in the same way that ParU relies on CHOLMOD.  Both use it for its ordering options (which doesn't use the GPU); they don't actually need the numerical factorization in CHOLMOD (which can be done on the GPU).

If this looks OK to you then we'll leave my most recent change to ParU as-is.

```
( cd ParU && make )
make[1]: Entering directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU'
( cd build && cmake  .. && cmake --build . --config Release -j40 )
-- Building PARU version: v0.1.0 (Dec 30, 2023)
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Source:        /home/faculty/d/davis/dev2/SuiteSparse/ParU 
-- Build:         /home/faculty/d/davis/dev2/SuiteSparse/ParU/build 
-- Install lib:      lib
-- Install include:  include
-- Install bin:      bin
-- Install pkg-file: lib
-- Install rpath:    
-- Build   rpath:    /home/faculty/d/davis/dev2/SuiteSparse/ParU/build
-- Build type:    Release 
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - /usr/bin/f95
-- The Fortran compiler identification is GNU 9.4.0
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/f95 - skipped
-- Fortran: /usr/bin/f95
-- CUDA: not enabled
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP_Fortran: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- SuiteSparse_config version: 7.4.0
-- SuiteSparse_config include: /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config
-- SuiteSparse_config library: /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config/build/libsuitesparseconfig.so.7.4.0
-- SuiteSparse_config static:  /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config/build/libsuitesparseconfig.a
-- AMD version: 3.3.0
-- AMD include: /home/faculty/d/davis/dev2/SuiteSparse/AMD/Include
-- AMD library: /home/faculty/d/davis/dev2/SuiteSparse/AMD/build/libamd.so.3.3.0
-- AMD static:  /home/faculty/d/davis/dev2/SuiteSparse/AMD/build/libamd.a
-- COLAMD version: 3.3.0
-- COLAMD include: /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/Include
-- COLAMD library: /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/build/libcolamd.so.3.3.0
-- COLAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/build/libcolamd.a
-- CAMD version: 3.3.0
-- CAMD include: /home/faculty/d/davis/dev2/SuiteSparse/CAMD/Include
-- CAMD library: /home/faculty/d/davis/dev2/SuiteSparse/CAMD/build/libcamd.so.3.3.0
-- CAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/CAMD/build/libcamd.a
-- CCOLAMD version: 3.3.0
-- CCOLAMD include: /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/Include
-- CCOLAMD library: /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/build/libccolamd.so.3.3.0
-- CCOLAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/build/libccolamd.a
-- CHOLMOD version: 5.1.0
-- CHOLMOD include: /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/Include
-- CHOLMOD library: /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/libcholmod.so.5.1.0
-- CHOLMOD static:  /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/libcholmod.a
-- UMFPACK version: 6.3.0
-- UMFPACK include: /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/Include
-- UMFPACK library: /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/build/libumfpack.so.6.3.0
-- UMFPACK static:  /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/build/libumfpack.a
-- Looking for Intel 32-bit BLAS
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - found
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libmkl_gf_lp64.so;/usr/lib/x86_64-linux-gnu/libmkl_gnu_thread.so;/usr/lib/x86_64-linux-gnu/libmkl_core.so;/usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;-lpthread;-lm;-ldl  
-- Found Intel10_64lp 32-bit BLAS
-- OpenMP C++ libraries:    /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;/usr/lib/x86_64-linux-gnu/libpthread.so 
-- OpenMP C++ include:       
-- OpenMP C++ flags:        -fopenmp 
-- BLAS libraries:      /usr/lib/x86_64-linux-gnu/libmkl_gf_lp64.so;/usr/lib/x86_64-linux-gnu/libmkl_gnu_thread.so;/usr/lib/x86_64-linux-gnu/libmkl_core.so;/usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;-lpthread;-lm;-ldl 
-- BLAS linker flags:    
-- Skipping the demos in ParU/Demo
-- ------------------------------------------------------------------------
-- SuiteSparse CMAKE report for: paru
-- ------------------------------------------------------------------------
-- inside common SuiteSparse root:  false
-- install in SuiteSparse/lib and SuiteSparse/include: OFF
-- build type:           Release
-- BUILD_SHARED_LIBS:    ON
-- BUILD_STATIC_LIBS:    ON
-- use OpenMP:           yes 
-- C compiler:           /usr/bin/cc 
-- C flags:              
-- C++ compiler:         /usr/bin/c++
-- C++ flags:            
-- C Flags release:      -O3 -DNDEBUG 
-- C++ Flags release:    -O3 -DNDEBUG 
-- Fortran compiler:     /usr/bin/f95 
-- compile definitions:  BLAS_Intel10_64lp
-- BLAS integer:         int32_t
-- ------------------------------------------------------------------------
-- Configuring done (3.5s)
CMake Error at CMakeLists.txt:250 (target_include_directories):
  Error evaluating generator expression:

    $<TARGET_PROPERTY:CUDA::cublas,INTERFACE_INCLUDE_DIRECTORIES>

  Target "CUDA::cublas" not found.


CMake Error at CMakeLists.txt:254 (target_include_directories):
  Error evaluating generator expression:

    $<TARGET_PROPERTY:CUDA::cublas,INTERFACE_INCLUDE_DIRECTORIES>

  Target "CUDA::cublas" not found.


CMake Error at /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/CHOLMODTargets.cmake:60 (set_target_properties):
  The link interface of target "SuiteSparse::CHOLMOD_static" contains:

    CUDA::nvrtc

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/CHOLMODConfig.cmake:155 (include)
  CMakeLists.txt:80 (find_package)


-- Generating done (0.0s)
CMake Generate step failed.  Build files cannot be regenerated correctly.
make[1]: *** [Makefile:38: library] Error 1
make[1]: Leaving directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU'
make: *** [Makefile:43: library] Error 2
hypersparse $ 
```

After https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/eb2809103b76ef907807a7c63c57038d26e5fb3a everything works:
```

( cd ParU && make )
make[1]: Entering directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU'
( cd build && cmake  .. && cmake --build . --config Release -j8 )
-- Building PARU version: v0.1.0 (Dec 30, 2023)
-- Source:        /home/faculty/d/davis/dev2/SuiteSparse/ParU 
-- Build:         /home/faculty/d/davis/dev2/SuiteSparse/ParU/build 
-- Install lib:      lib
-- Install include:  include
-- Install bin:      bin
-- Install pkg-file: lib
-- Install rpath:    
-- Build   rpath:    /home/faculty/d/davis/dev2/SuiteSparse/ParU/build
-- Build type:    Release 
-- Fortran: /usr/bin/f95
-- Looking for a CUDA compiler
-- Looking for a CUDA compiler - /usr/local/cuda/bin/nvcc
-- Looking for CUDA
-- Find CUDA tool kit:
-- Found CUDAToolkit: /usr/local/cuda/include (found version "11.7.99") 
-- CUDA toolkit found:   TRUE
-- CUDA toolkit version: 11.7.99
-- CUDA toolkit include: /usr/local/cuda/include
-- CUDA toolkit lib dir: /usr/local/cuda/lib64
-- The CUDA compiler identification is NVIDIA 11.7.99
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- CUDA: enabled
-- SuiteSparse_config version: 7.4.0
-- SuiteSparse_config include: /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config
-- SuiteSparse_config library: /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config/build/libsuitesparseconfig.so.7.4.0
-- SuiteSparse_config static:  /home/faculty/d/davis/dev2/SuiteSparse/SuiteSparse_config/build/libsuitesparseconfig.a
-- AMD version: 3.3.0
-- AMD include: /home/faculty/d/davis/dev2/SuiteSparse/AMD/Include
-- AMD library: /home/faculty/d/davis/dev2/SuiteSparse/AMD/build/libamd.so.3.3.0
-- AMD static:  /home/faculty/d/davis/dev2/SuiteSparse/AMD/build/libamd.a
-- COLAMD version: 3.3.0
-- COLAMD include: /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/Include
-- COLAMD library: /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/build/libcolamd.so.3.3.0
-- COLAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/COLAMD/build/libcolamd.a
-- CAMD version: 3.3.0
-- CAMD include: /home/faculty/d/davis/dev2/SuiteSparse/CAMD/Include
-- CAMD library: /home/faculty/d/davis/dev2/SuiteSparse/CAMD/build/libcamd.so.3.3.0
-- CAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/CAMD/build/libcamd.a
-- CCOLAMD version: 3.3.0
-- CCOLAMD include: /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/Include
-- CCOLAMD library: /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/build/libccolamd.so.3.3.0
-- CCOLAMD static:  /home/faculty/d/davis/dev2/SuiteSparse/CCOLAMD/build/libccolamd.a
-- CHOLMOD version: 5.1.0
-- CHOLMOD include: /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/Include
-- CHOLMOD library: /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/libcholmod.so.5.1.0
-- CHOLMOD static:  /home/faculty/d/davis/dev2/SuiteSparse/CHOLMOD/build/libcholmod.a
-- UMFPACK version: 6.3.0
-- UMFPACK include: /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/Include
-- UMFPACK library: /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/build/libumfpack.so.6.3.0
-- UMFPACK static:  /home/faculty/d/davis/dev2/SuiteSparse/UMFPACK/build/libumfpack.a
-- Looking for Intel 32-bit BLAS
-- Found Intel10_64lp 32-bit BLAS
-- OpenMP C++ libraries:    /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;/usr/lib/x86_64-linux-gnu/libpthread.so 
-- OpenMP C++ include:       
-- OpenMP C++ flags:        -fopenmp 
-- BLAS libraries:      /usr/lib/x86_64-linux-gnu/libmkl_gf_lp64.so;/usr/lib/x86_64-linux-gnu/libmkl_gnu_thread.so;/usr/lib/x86_64-linux-gnu/libmkl_core.so;/usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so;-lpthread;-lm;-ldl 
-- BLAS linker flags:    
-- Skipping the demos in ParU/Demo
-- ------------------------------------------------------------------------
-- SuiteSparse CMAKE report for: paru
-- ------------------------------------------------------------------------
-- inside common SuiteSparse root:  false
-- install in SuiteSparse/lib and SuiteSparse/include: OFF
-- build type:           Release
-- BUILD_SHARED_LIBS:    ON
-- BUILD_STATIC_LIBS:    ON
-- use OpenMP:           yes 
-- C compiler:           /usr/bin/cc 
-- C flags:              
-- C++ compiler:         /usr/bin/c++
-- C++ flags:            
-- C Flags release:      -O3 -DNDEBUG 
-- C++ Flags release:    -O3 -DNDEBUG 
-- Fortran compiler:     /usr/bin/f95 
-- compile definitions:  BLAS_Intel10_64lp
-- BLAS integer:         int32_t
-- CUDA architectures:   52;75;80
-- ------------------------------------------------------------------------
-- Configuring done (5.4s)
-- Generating done (0.0s)
-- Build files have been written to: /home/faculty/d/davis/dev2/SuiteSparse/ParU/build
gmake[2]: Entering directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU/build'
[  1%] Building CXX object CMakeFiles/ParU.dir/Source/paru_Diag_update.cpp.o
[  2%] Building CXX object CMakeFiles/ParU.dir/Source/paru_analyze.cpp.o
[  3%] Building CXX object CMakeFiles/ParU_static.dir/Source/paru_Diag_update.cpp.o
[  4%] Building CXX object CMakeFiles/ParU.dir/Source/paru_assemble.cpp.o
[  7%] Building CXX object CMakeFiles/ParU_static.dir/Source/paru_assemble.cpp.o
[  7%] Building CXX object CMakeFiles/ParU_static.dir/Source/paru_analyze.cpp.o
[  8%] Building CXX object CMakeFiles/ParU.dir/Source/paru_assemble_row2U.cpp.o

...
[ 96%] Building CXX object CMakeFiles/ParU_static.dir/Source/paru_version.cpp.o
[ 97%] Building CXX object CMakeFiles/ParU_static.dir/Source/paru_write.cpp.o
[ 98%] Linking CXX shared library libparu.so
[ 98%] Built target ParU
[100%] Linking CXX static library libparu.a
[100%] Built target ParU_static
gmake[2]: Leaving directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU/build'
make[1]: Leaving directory '/home/faculty/d/davis/dev2/SuiteSparse/ParU'
```

and the demos work in ParU.

--- a/CHOLMOD/Config/CHOLMODConfig.cmake.in
+++ b/CHOLMOD/Config/CHOLMODConfig.cmake.in
@@ -38,6 +38,19 @@ set ( CHOLMOD_VERSION "@CHOLMOD_VERSION_MAJOR@.@CHOLMOD_VERSION_MINOR@.@CHOLMOD_
 include ( CMakeFindDependencyMacro )
 set ( _dependencies_found ON )
 
+# Look for NVIDIA CUDA toolkit
+if ( @SUITESPARSE_CUDA@ AND NOT CUDAToolkit_FOUND )
+    find_dependency ( CUDAToolkit @CUDAToolkit_VERSION_MAJOR@ )
+    if ( NOT CUDAToolkit_FOUND )
+        set ( _dependencies_found OFF )
+    endif ( )
+endif ( )
+
+if ( NOT _dependencies_found )
+    set ( CHOLMOD_FOUND OFF )
+    return ( )
+endif ( )
+
 # Look for SuiteSparse_config, COLAMD, and AMD targets
 if ( @SUITESPARSE_IN_BUILD_TREE@ )
     if ( NOT TARGET SuiteSparse::SuiteSparseConfig )

--- a/ParU/CMakeLists.txt
+++ b/ParU/CMakeLists.txt
@@ -38,14 +38,7 @@ set ( CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH}
     ${PROJECT_SOURCE_DIR}/../SuiteSparse_config/cmake_modules
     )
 
-option ( ENABLE_CUDA "Enable CUDA acceleration" on )
-
 include ( SuiteSparsePolicy )
-
-if ( SUITESPARSE_CUDA )
-    # ParU with CHOLMOD (which can use CUDA)
-    enable_language ( CUDA )
-endif ( )
 
 #-------------------------------------------------------------------------------
 # find library dependencies

--- a/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake
+++ b/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake
@@ -16,12 +16,9 @@
 #
 #   ENABLE_CUDA:        if set to true, CUDA is enabled for the project.
 #                       Default: true for CHOLMOD and SPQR, which use the GPU
-#                       for their numerical factorizsation.  It is also enabled
-#                       for UMFPACK, KLU, and ParU (which use CHOLMOD and this
-#                       require this flag if CHOLMOD has been compiled with
-#                       ENABLE_CUDA).  The flag is false for false for
-#                       GraphBLAS since CUDA for that package is in progress
-#                       and not ready for production use.
+#                       for their numerical factorizsation.  The flag is false
+#                       for GraphBLAS since CUDA for that package is in
+#                       progress and not ready for production use.
 #
 #   LOCAL_INSTALL:      if true, "cmake --install" will install
 #                       into SuiteSparse/lib and SuiteSparse/include.

